internal/api/shared/schemas/schemas.go

```diff
@@ -82,65 +94,102 @@ func New(conf config.Schemas) Schemas {
 }
 
 // ValidatePanels verify a list of panels.
 // The panels are matched against the known list of CUE definitions (schemas).
 // If no schema matches for at least 1 panel, the validation fails.
 func (s *sch) ValidatePanels(panels map[string]*modelV1.Panel) error {
+	if s.panels == nil {
+		logrus.Warning("panel schemas are not loaded")
+		return nil
+	}
 	// go through the panels list
 	// the processing stops as soon as it detects an invalid panel -> TODO: improve this to return a list of all the errors encountered ?
 	for panelName, panel := range panels {
 		logrus.Tracef("Panel to validate: %s", panelName)
-		panelPluginByte, err := panel.Spec.Plugin.JSONMarshal()
-		if err != nil {
-			logrus.WithError(err).Debugf("unable to marshal the panel plugin %q", panel.Spec.Plugin.Kind)
+		if err := s.validatePlugin(panel.Spec.Plugin, "panel", panelName, s.panels, func(originalValue cue.Value) cue.Value {
+			if s.queries != nil {
+				return originalValue.Unify(s.queries.disjSchema)
+			}
+			return originalValue
+		}); err != nil {
 			return err
 		}
-		// compile the JSON panel plugin into a CUE Value
-		value := s.context.CompileBytes(panelPluginByte)
+	}
+	logrus.Debug("All panels are valid")
+	return nil
+}
 
-		// retrieve the corresponding panel schema
-		panelSchema, err := retrieveSchemaForKind(panelName, value, s.panels.kindCuePath, s.panels.schemas)
-		if err != nil {
-			return err
-		}
+func (s *sch) validatePlugin(plugin modelV1.Plugin, modelKind string, modelName string, cueDefs *cueDefs, enrichSchema func(originalValue cue.Value) cue.Value) error {
+	pluginData, err := plugin.JSONMarshal()
+	if err != nil {
+		logrus.WithError(err).Debugf("unable to marshal the plugin %q", plugin.Kind)
+		return err
+	}
+	// compile the JSON plugin into a CUE Value
+	value := s.context.CompileBytes(pluginData)
 
-		// Then merge with the queries disjunction schema
-		panelSchema = panelSchema.Unify(s.queries.disjSchema)
-		if panelSchema.Err() != nil {
-			logrus.WithError(panelSchema.Err()).Error("Error unifying panel schema with queries schema")
-			return panelSchema.Err()
-		}
+	// retrieve the corresponding schema
+	pluginSchema, err := retrieveSchemaForKind(modelKind, modelName, value, cueDefs.kindCuePath, cueDefs.schemas)
+	if err != nil {
+		return err
+	}
 
-		// do the validation using the main #panel def of the schema as entrypoint
-		unified := value.Unify(panelSchema.LookupPath(cue.ParsePath("")))
-		opts := []cue.Option{
-			cue.Concrete(true),
-			cue.Attributes(true),
-			cue.Definitions(true),
-			cue.Hidden(true),
-		}
-		err = unified.Validate(opts...)
-		if err != nil {
-			logrus.Debug(errors.Details(err, nil))
-			//TODO: return errors.Details(err, nil) to get a more meaningful error, but should be cleaned of line numbers & server file paths!
-			err = fmt.Errorf("invalid panel %s: %s", panelName, err) // enrich the error message returned by cue lib
-			return err
-		}
+	// Then merge with the queries disjunction schema
```

Comment:
since this function is now more generic/abstracted, this comment should be moved or removed (dosnt apply to datasources validation)